src/transformStateWithClones.js

```diff
@@ -7,7 +7,43 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const STATECOPY = {};
+
+  Object.assign(STATECOPY, state);
+
+  const modificationsHistory = [];
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      Object.assign(STATECOPY, action.extraData);
+
+      const STATESARRAY = { ...STATECOPY };
+
+      modificationsHistory.push(STATESARRAY);
+    }
+
+    if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete STATECOPY[key];
+      }
+
+      const STATESARRAY = { ...STATECOPY };
+
+      modificationsHistory.push(STATESARRAY);
+    }
+
+    if (action.type === 'clear') {
+      for (const key in STATECOPY) {
+        delete STATECOPY[key];
+      }
+
+      const STATESARRAY = { ...STATECOPY };
+
+      modificationsHistory.push(STATESARRAY);
+    }
+  }
```

Comment:
Instead of using multiple `if` statements, a `switch` statement would be more appropriate here since you are checking against the `action.type`. Also, remember to include a `default` case to handle any unexpected action types.